Mathlib/Data/Vector3.lean

```diff
@@ -0,0 +1,330 @@
+/-
+Copyright (c) 2017 Mario Carneiro. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Mario Carneiro
+
+! This file was ported from Lean 3 source module data.vector3
+! leanprover-community/mathlib commit 3d7987cda72abc473c7cdbbb075170e9ac620042
+! Please do not edit these lines, except to modify the commit id
+! if you have ported upstream changes.
+-/
+import Mathlib.Data.Fin.Fin2
+import Mathlib.Init.Align
+import Mathlib.Mathport.Notation
+
+/-!
+# Alternate definition of `Vector` in terms of `Fin2`
+
+This file provides a locale `Vector3` which overrides the `[a, b, c]` notation to create a `Vector3`
+instead of a `List`.
+
+The `::` notation is also overloaded by this file to mean `Vector3.cons`.
+-/
+
+open Fin2 Nat
+
+universe u
+
+variable {α : Type _} {m n : ℕ}
+
+/-- Alternate definition of `Vector` based on `Fin2`. -/
+def Vector3 (α : Type u) (n : ℕ) : Type u :=
+  Fin2 n → α
+#align vector3 Vector3
+
+instance [Inhabited α] : Inhabited (Vector3 α n) where
+  default := fun _ => default
+
+namespace Vector3
+
+/-- The empty vector -/
+@[match_pattern]
+def nil : Vector3 α 0 :=
+  fun.
+#align vector3.nil Vector3.nil
+
+/-- The vector cons operation -/
+@[match_pattern]
+def cons (a : α) (v : Vector3 α n) : Vector3 α (succ n) := fun i => by
+  refine' i.cases' _ _
+  exact a
+  exact v
+#align vector3.cons Vector3.cons
+
+-- Porting note: was
+-- scoped notation3 "["(l", "* => foldr (h t => cons h t) nil)"]" => l
+
+section
+open Lean
+
+-- this is copied from `src/Init/Notation.lean`
+scoped macro_rules
+  | `([ $elems,* ]) => do
+    -- NOTE: we do not have `TSepArray.getElems` yet at this point
+    let rec expandLit (i : Nat) (skip : Bool) (result : TSyntax `term) : MacroM Syntax := do
+      match i, skip with
+      | 0,   _     => pure result
+      | i+1, true  => expandLit i false result
+      | i+1, false => expandLit i true (←``(Vector3.cons $(⟨elems.elemsAndSeps.get! i⟩) $result))
+    if elems.elemsAndSeps.size < 64 then
+      expandLit elems.elemsAndSeps.size false (← ``(Vector3.nil))
+    else
+      `(%[ $elems,* | Vector3.nil ])
+
+-- this is copied from `src/Init/NotationExtra.lean`
+scoped macro_rules
+  | `(%[ $[$x],* | $k ]) =>
+    if x.size < 8 then
+      x.foldrM (β := Term) (init := k) fun x k =>
+        `(Vector3.cons $x $k)
+    else
+      let m := x.size / 2
+      let y := x[m:]
+      let z := x[:m]
+      `(let y := %[ $[$y],* | $k ]
+        %[ $[$z],* | y ])
+
+-- this is copied from `src/Init/NotationExtra.lean`
+@[app_unexpander Vector3.nil] def unexpandNil : Lean.PrettyPrinter.Unexpander
+  | `($(_)) => `([])
+
+-- this is copied from `src/Init/NotationExtra.lean`
+@[app_unexpander Vector3.cons] def unexpandCons : Lean.PrettyPrinter.Unexpander
+  | `($(_) $x [])      => `([$x])
+  | `($(_) $x [$xs,*]) => `([$x, $xs,*])
+  | _                  => throw ()
+
+end
```

Comment:
That macro in `Init.Notation` is actually the source of a lot of problems and I wouldn't recommend copying it. I don't think we'll be using this for long vectors anyway so I would just do the simple thing. The reason why the `notation3` doesn't work is because it creates an overloading `syntax`, which is presumably why the replacement is using `macro_rules` instead of `macro`. We can do something a lot simpler as the `macro_rules` analogue of `notation3`:

```lean
scoped macro_rules | `([$l,*]) => `(expand_foldr% (h t => cons h t) nil [$(.ofElems l),*])
```
This replaces the two macro_rules, but not the unexpanders.